pandas/core/generic.py

```diff
@@ -1995,10 +1996,11 @@ def empty(self) -> bool_t:
     def __array__(self, dtype: npt.DTypeLike | None = None) -> np.ndarray:
         values = self._values
         arr = np.asarray(values, dtype=dtype)
-        if arr is values and using_copy_on_write():
-            # TODO(CoW) also properly handle extension dtypes
-            arr = arr.view()
-            arr.flags.writeable = False
+        if arr is values and using_copy_on_write() and self._mgr.is_single_block:
+            # Check if self._values coerced data
+            if not is_1d_only_ea_dtype(self.dtypes.iloc[0]):
```

Comment:
It is possible to not have a first column? And do we never have any block in a case of a DataFrame of zero columns?

Reply:
Also, does `is_1d_only_ea_dtype` guarantee it will always have coerced the data? I suppose a general EA can give you a view when converting it to a numpy array (for .values it only needs a reshape to 2D, but that's a view)

Reply:
Ah thanks very much. I was too focused on integers where we are coercing to object right now which causes a copy. With dtypes where converting to object does not cause a copy this still shares memory of course. 

Adjusted the check

Reply:
If we have zero columns then we have an empty array, correct? We can only modify an empty array through enlarging the array which would cause a copy anyway? Am I missing something?

Reply:
Ah I misunderstood you yesterday then. With empty you mean no columns, otherwise iloc would work, correct? In this case the is_single_block check fails and hence we don't get there. I'll add a test for this though, because we could easily cause a regression if we are not careful

Reply:
> In this case the is_single_block check fails and hence we don't get there

Yes, but so I was also wondering if it would be possible to have a zero-column dataframe with one block (so whether it's possible to have a block with an axis of size 0). 

It seems to be theoretically possible, by constructing this manually:

```
In [82]: block = pd.core.internals.make_block(np.zeros((0, 10)), np.array([]))

In [86]: mgr = pd.core.internals.BlockManager([block], [pd.Index([]), pd.Index(range(10))])

In [88]: df = pd.DataFrame(mgr)

In [89]: df
Out[89]: 
Empty DataFrame
Columns: []
Index: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [90]: df._mgr.is_single_block
Out[90]: True
```

But I don't know if there is any way you could get that through actual pandas operations.

Reply:
I don't think so. It should be a bug if you could do this?